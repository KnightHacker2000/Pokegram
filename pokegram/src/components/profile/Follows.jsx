import PropTypes from 'prop-types';
import * as React from 'react';
import {
  useEffect, useRef, useState
} from 'react';
import {
  Box, Container
} from '@mui/material';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import Divider from '@mui/material/Divider';
import ListItemText from '@mui/material/ListItemText';
import ListItemAvatar from '@mui/material/ListItemAvatar';
import Avatar from '@mui/material/Avatar';
import ClearIcon from '@mui/icons-material/Clear';
import userService from '../../services/userService';
import HomeState from '../../models/homeState';

function Follows(props) {
  const {
    showSug, UID, handleShowFo, homeStates
  } = props;
  console.log(UID);
  const [, setFoList] = useState(false);
  const [sugObj, setSugObj] = useState({ id: [1], usernames: [2], foList: [3] });

  const firstRendering = useRef(true);
  useEffect(() => {
    async function fetchFoSugbyUID(id) {
      let params = `{"userId":${id} }`;
      const foList = await userService.getUserById(JSON.parse(params));
      params = `{"id":${id} }`;
      const arr = [];
      await userService.getFoSug(JSON.parse(params)).then((data) => {
        data.users.forEach(async (user) => {
          params = `{"userId":${user} }`;
          const tmp = await userService.getUserById(JSON.parse(params));
          arr.push(tmp.username);
        });
        setSugObj({ id: data.users, usernames: arr, foList: foList.follows });
      });
    }

    if (firstRendering.current) {
      firstRendering.current = false;
      // hardcoded UID, use UID prop in the fut
      fetchFoSugbyUID(2);
      setFoList(true);
    }
  });

  const handleClear = () => {
    handleShowFo();
  };

  const handleClick = (event) => {
    event.preventDefault();
    const otherUID = event.currentTarget.getAttribute('data-index');
    homeStates.handleHomeStates(false, true, false, false, false, otherUID);
    handleClear();
  };

  const foSug = (s) => {
    const prompt = <p>Follower Suggestions For You:</p>;
    let listing;
    if (s.id.length === 0) {
      listing = <p>No Suggestions</p>;
    } else {
      listing = s.id.map((ele, i) => (
        <ListItem key={Math.random()}>
          <ListItemAvatar key={Math.random()} onClick={handleClick} data-index={ele}>
            <Avatar alt={s.usernames[i]} src="/static/images/avatar/3.jpg" />
          </ListItemAvatar>
          <ListItemText
            primary={ele}
          />
        </ListItem>
      ));
    }
    return (
      <div>
        {prompt}
        {listing}
      </div>
    );
  };

  const foList = () => {
    const prompt = <p>Following: </p>;
    const listing = sugObj.foList.map((ele) => (
      <ListItem key={ele}>
        <ListItemAvatar>
          <Avatar alt={ele.toString()} src="/static/images/avatar/3.jpg" />
        </ListItemAvatar>
        <ListItemText
          primary={ele}
        />
      </ListItem>
    ));
    return (
      <div>
        {prompt}
        {listing}
      </div>
    );
  };

  return (
    <div style={{
      position: 'fixed',
      top: 0,
      bottom: 0,
      left: 0,
      right: 0,
      display: 'flex',
      alignItems: 'center',
      flexDirection: 'row',
      background: 'rgba(90, 90, 90, 0.8)',
      zIndex: 99
    }}
    >
      <div style={{
        zIndex: 99,
        display: 'flex',
        marginLeft: 'auto',
        marginRight: 'auto',
        marginTop: 'auto',
        marginBottom: 'auto',
        alignItems: 'center',
        background: 'white',
        border: '1px solid rgba(0, 0, 0, 0.05)',
        borderRadius: '10px',
        height: 'max-content'
      }}
      >
        <Container maxWidth="xs">
          <Box sx={{ display: 'flex', alignItems: 'center' }}>
            <ClearIcon sx={{ width: '200%' }} onClick={handleClear} />
          </Box>
          <List sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}>
            {showSug && sugObj.foList.length > 0 && foSug(sugObj)}
            <Divider />
            {foList()}
          </List>
        </Container>
      </div>
    </div>
  );
}

Follows.propTypes = {
  showSug: PropTypes.bool,
  UID: PropTypes.number,
  handleShowFo: PropTypes.func,
  homeStates: PropTypes.instanceOf(HomeState)
};

Follows.defaultProps = {
  showSug: false,
  UID: -1,
  handleShowFo: null,
  homeStates: null
};

export default Follows;
