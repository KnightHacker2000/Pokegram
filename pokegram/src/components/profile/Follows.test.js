/**
 * @jest-environment jsdom
 */
import React from 'react';
//  import { fireEvent, render } from '@testing-library/react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Follows from './Follows';
import HomeState from '../../models/homeState';
import '@testing-library/jest-dom/extend-expect';
// import { test } from '@jest/globals';

const parentStates = new HomeState(
  2,
  (isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID) => {
    console.log(isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID);
  },
  false,
  false,
  true,
  false,
  2
);

test('test follow snapshot', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Follows showSug={true} UID={2} handleShowFo={()=>{}} homeStates={parentStates} />);
  const followComp = renderer.getRenderOutput();
  // expect(tree).toMatchSnapshot();
});

// test('test profile click', () => {
//   const renderer = new ShallowRenderer();
//   renderer.render(<ResponsiveAppBar homeStates={parentStates} />);
//   const res = renderer.getRenderOutput();
//   // const sha = shallow()
//   // const tree = menuComp.toJSON();
//   res.props.children.props.children.props.children[0].props.onClick();
//   // expect(tree).toMatchSnapshot();
// });

// test('test follow button', () => {
//   const renderer = new ShallowRenderer();
//   renderer.render(<Profile homeStates={parentStates} />);
//   const res = renderer.getRenderOutput();
//   res.props.children[0].props.children.props.children[1].props.children[0].props.children[1].props.onClick({ preventDefault: () => {} });
// });