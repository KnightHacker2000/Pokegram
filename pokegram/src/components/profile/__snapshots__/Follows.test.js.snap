// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test fo sug 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "background": "rgba(90, 90, 90, 0.8)",
      "bottom": 0,
      "display": "flex",
      "flexDirection": "row",
      "left": 0,
      "position": "fixed",
      "right": 0,
      "top": 0,
      "zIndex": 99,
    }
  }
>
  <div
    style={
      {
        "alignItems": "center",
        "background": "white",
        "border": "1px solid rgba(0, 0, 0, 0.05)",
        "borderRadius": "10px",
        "display": "flex",
        "height": "max-content",
        "marginBottom": "auto",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": "auto",
        "zIndex": 99,
      }
    }
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
    >
      <div
        className="MuiBox-root css-70qvj9"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15wpu8i-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
      <ul
        className="MuiList-root MuiList-padding css-1mcnwpj-MuiList-root"
      >
        <div>
          <p>
            Follower Suggestions For You:
          </p>
          <li
            className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            disabled={false}
          >
            <div
              className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              data-index={1}
              onClick={[Function]}
            >
              <div
                className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
              >
                <img
                  alt={2}
                  className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="/static/images/avatar/3.jpg"
                />
              </div>
            </div>
            <div
              className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                1
              </span>
            </div>
          </li>
        </div>
        <hr
          className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div>
          <p>
            Following: 
          </p>
          <li
            className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            disabled={false}
          >
            <div
              className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
            >
              <div
                className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
              >
                <img
                  alt="3"
                  className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="/static/images/avatar/3.jpg"
                />
              </div>
            </div>
            <div
              className="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                3
              </span>
            </div>
          </li>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
